### PR TITLE
Feature/32/screenplay

### LINKED ISSUE #32: Error: Wrong script generation
- Script is being generated from available cards rather than game's selected cards

### PR DESCRIPTION
**Pull request by mamodom**
closes https://github.com/lycan-city/werewolf-moderator/issues/32

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,6 +1,3 @@
 language: node_js
 node_js:
   - "7"
-  - "6"
```
Comment (vetom) [inline .travis.yml:4]: :eyes:

**Review by vetom — APPROVED** [review]

File: .travis.yml (hunk 1 above)
Comment (mamodom) [reply .travis.yml:4]: I'd like @wOvalle's opinion on this

Comment (wovalle) [reply .travis.yml:4]: My opinion is: it depends.

Are we transpiling the code anyways? When someone downloads our npm package it uses the transpiled code?

If both of those questions are correct then yes.

File: tests/game.js [hunk 2]
```diff
@@ -13,65 +14,77 @@ const cards = [
...
+    describe('Create', () => {
+        describe('given mode is NORMAL', () => {
+            beforeEach(function () {
+                this.game = game.create(players, language, cards, game.mode.NORMAL);
+            });
 
-    it('should contain the weight on each normal game created.', () => {
-        normalGame.should.have.property('weight').which.is.a.Number();
-    });
+            it('should create a normal game.', function () {
```
Comment (wovalle) [inline tests/game.js:34]: I know there was a limitation of using arrow functions in mocha's tests but I can't remember it. This was the reason behind your refactor changing the functions?

File: tests/game.js [hunk 3]
```diff
@@ -13,65 +14,77 @@ const cards = [
...
-    });
+            it('should includes the maximun number of players on each normal game created.', function () {
+                this.game.players.should.be.equal(players);
+            });
+        });
 
-    it('should give a weight within -20 and 20 on each chaos game created.', () => {
-        chaosGame.weight.should.be.within(-20, 20);
-    });
+        describe('given mode is NORMAL', () => {
```
Comment (wovalle) [inline tests/game.js]: `given mode is CHAOS`

File: tests/screenplay.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                .and.to.not.be.null;
+        });
+
+        it('is expected to be an array', function () {
+            expect(this.screenplay).to.be.an('array');
+        });
+
+        it('is expected to have `key` and `level`', function () {
+            this.screenplay.forEach(call =>
+                expect(call).to.have.all.keys('key', 'level', 'firstNight')
```
Comment (wovalle) [inline tests/screenplay.js:42]: Didn't know about `.to.have.all.keys`. good one!

File: tests/screenplay.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
+/* eslint-disable no-undef*/
+/* eslint-disable no-unused-expressions*/
+/* eslint-disable func-names */
+const chai = require('chai');
+const moderator = require('../src/moderator');
+
+const expect = chai.expect;
+
+describe('Screenplay', () => {
+    describe('fromGame()', () => {
```
Comment (wovalle) [inline tests/screenplay.js:10]: I would add tests for each specific filter scenario.

`witch` should have all level
`cult_leader` should have deck level
`villager` should have game

Don't you think is necessary?

File: .travis.yml (hunk 1 above)
Comment (mamodom) [reply .travis.yml:4]: Yes, we are transpiling and bundling the source you can see [here](https://github.com/lycan-city/werewolf-brain/blob/master/package.json#L5) that we're not exposing the es6 version

File: tests/game.js (hunk 2 above)
Comment (mamodom) [reply tests/game.js:34]: see [this](https://github.com/lycan-city/werewolf-brain/pull/77/commits/65c6737be473819b78c3d25d89a88babe15c58ce) commit message.

By using arrow functions there is no context

File: tests/game.js (hunk 3 above)
Comment (mamodom) [reply tests/game.js]: Nice catch

File: tests/screenplay.js (hunk 4 above)
Comment (mamodom) [reply tests/screenplay.js:42]: That's one of the reasons I installed chai

File: tests/screenplay.js (hunk 5 above)
Comment (mamodom) [reply tests/screenplay.js:10]: Something like that was my original intent but there's a catch:

`witch`, `cult_leader`, and `villager` are roleKeys, not callKeys.

In this particular example, the roles `cult_leader` and `witch` are bound each to one call whose key is the same roleKey, but in the case of some roles like `mason`, `cupid`, `fruit_brute`, and some other wolves they are not only  not bound to a single call, but their calls don't necessarily match their respective roleKeys.

To be able to test the actual values (and not only the structure) of the resulting screenplay, and have a test that's not insanely brittle, I'd have to mock or inject or whatever else to be able to properly assert upon those values provided on the test and not on something foreign to the test.

**Pull request merged**